privateSBOMExchange/src/petra/lib/models/__init__.py

```diff
@@ -1,13 +1,304 @@
+            A list of FieldNode instances representing the fields of the package.
+        """
+        self.metadata_type_name:str= type_name  # Store data as string
+        self.metadata_type_value:str=type_value
+        self.encrypted_data=None #store encrypted data when visited by Encrypt visitor
+        self.hash=None #store hashed data when visited by Merkle visitor
+        self.children = children
+
+    def accept(self, visitor):
+        return visitor.visit_complex_node(self)
+
+
+class SbomNode(Node):
+    """Represents the root node of a Software Bill of Materials (SBOM) tree.
+
+    This node can have three types of children, all of which are derived from the Node class:
+    1. **FieldNode**: Represents a field in an SBOM document.
+    2. **ComplexNode**: Represents a complex node that has multiple fields.
+    3. **SbomNode**: Represents a package dependency that does have an available SBOM tree in the database.
+
+    Attributes
+    ----------
+    SBOM_name : str
+        The name or identifier of the SBOM.
+    encrypted_data : bytes or None
+        The encrypted verison of the data, initially set to None.
+    hash : bytes or None
+        The hashed version of the data, initially set to None.
+    children : List[Node]
+        A list of child nodes, which can be FieldNode, PackageNode, or other SbomNode instances.
+    """
+    def __init__(self,name:str,purl, children:List[Node]):
+        """Initialize an SbomNode.
+
+        Parameters
+        ----------
+        name : str
+            The name of the SBOM.
+        children : List[Node]
+            A list of child nodes that can be FieldNode, PackageNode, or other SbomNode instances.
+        """
+        self.SBOM_name=name
+        self.encrypted_data=None #store encrypted data when visited by Encrypt visitor
+        self.hash=None #store hashed data when visited by Merkle visitor
+        self.children = children  # Hold the children directly
+        self.signiture=None
+        self.pURL=purl
+    def accept(self, visitor):
+        # Accept the visitor on the root node and then on all children
+        return visitor.visit_sbom_node(self)
+
+  
+class MerkleVisitor:
+    """Visitor that computes the hash of the data in the nodes."""
+    def visit_field_node(self, node:FieldNode):
+        """Visit a FieldNode and compute the hash of its data.
+
+        The hash is computed using the formula:
+        if data is encrypted: H(node.encrypted_data)
+        If data is not encrypted: H("Field"|field_name|field_value)
+
+        Parameters
+        ----------
+        node : FieldNode
+            The field node whose data will be hashed.
+
+        Returns
+        -------
+        bytes
+            The computed hash of the field node data.
+        """
+        if node.encrypted_data is not None:
+            data_to_hash= node.encrypted_data
+        else:
+            data_to_hash=(f"Field{node.field_name}{node.field_value}").encode()
```

Comment:
I think we need to hash all of the data fields in a FieldNode, regardless of whether or not we have encrypted data. This is because we want to be be able to ensure the integrity of the whole node, not just individual fields. The same goes for the other node types.

This will also help us verify that the redaction was actually performed correctly: i.e., if we have encrypted data, we should only have placeholder values in the plain data variables. If we don't hash all of the node variables together, we can't check that this property holds.

Reply:
I'm ok with merging this PR without fixing this, but let's track this in an issue and make these changes in a future PR.

Reply:
Plus on this

Reply:
so encrypt all node attributes but children?

Reply:
I just opened  ticket #23 for this

Reply:
@SantiagoTorres